### PR TITLE
Improvements and cleanups for connecting to kopia server

### PR DESCRIPTION
**Pull request by jkowalski**
I was manually testing the robustness of the user experience of connecting to kopia repository server. I tried emulating mistakes users might make along the way and verifying that the UX makes sense.

* The biggest issue I've noticed that we were not properly setting up cache when `kopia repository connect server` was used.
* Another notable change is that `kopia cache` commands will be allowed for server-based repositories.
* ~~Also tightened validation of usernames to enforce that they must be lowercase and matching some sane regex.~~ (moved to #873)
* `KOPIA_CACHE_DIRECTORY` environment variable can now be used to globally override the cache path for kopia (both for create/connect and when opening the repository)
* Also made some errors when connecting to the server more descriptive and actually include useful information.

### CODE REVIEW

File: repo/local_config.go [hunk 1]
```diff
@@ -88,26 +90,34 @@ type repositoryObjectFormat struct {
...
+		}
+	}
 
-// Save writes the configuration to the specified writer.
-func (lc *LocalConfig) Save(w io.Writer) error {
-	b, err := json.MarshalIndent(lc, "", "  ")
+	b, err := json.MarshalIndent(lc2, "", "  ")
 	if err != nil {
-		return nil
+		return errors.Wrap(err, "error writing config file")
```
Comment (julio-lopez) [inline repo/local_config.go:109]: nit: _"error creating config file contents"_ or something to differentiate it from the actual `write` operation.

File: repo/local_config.go [hunk 2]
```diff
@@ -88,26 +90,34 @@ type repositoryObjectFormat struct {
...
 	}
 
-	_, err = w.Write(b)
+	if err = os.MkdirAll(filepath.Dir(filename), 0o700); err != nil {
+		return errors.Wrap(err, "unable to create config directory")
+	}
 
-	return errors.Wrap(err, "error saving local config")
+	return atomicfile.Write(filename, bytes.NewReader(b))
```
Comment (julio-lopez) [inline repo/local_config.go:116]: Should the error be wrapped?
`errors.Wrap` does the right thing when `err == nil`

File: repo/grpc_repository_client.go [hunk 3]
```diff
@@ -586,10 +591,17 @@ func (r *grpcRepositoryClient) UpdateDescription(d string) {
 }
 
 func (r *grpcRepositoryClient) Close(ctx context.Context) error {
+	if r.omgr == nil {
```
Comment (julio-lopez) [inline repo/grpc_repository_client.go:594]: is this safe with concurrency?

File: repo/api_server_repository.go [hunk 4]
```diff
@@ -204,8 +208,17 @@ func (r *apiServerRepository) UpdateDescription(d string) {
 }
 
 func (r *apiServerRepository) Close(ctx context.Context) error {
-	if err := r.omgr.Close(); err != nil {
-		return errors.Wrap(err, "error closing object manager")
+	if r.omgr != nil {
```
Comment (julio-lopez) [inline repo/api_server_repository.go:211]: Is `Close()` expected to be called concurrently?

File: repo/grpc_repository_client.go (hunk 3 above)
Comment (jkowalski) [reply repo/grpc_repository_client.go:594]: Close() is meant to be called without concurrency

File: repo/api_server_repository.go (hunk 4 above)
Comment (jkowalski) [reply repo/api_server_repository.go:211]: Close() is meant to be called without concurrency

**Pull request merged**